server/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,347 @@
+package resourcemanager
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/kubeshop/tracetest/server/id"
+	"github.com/mitchellh/mapstructure"
+)
+
+type ResourceSpec interface {
+	HasID() bool
+	Validate() error
+}
+
+type ResourceList[T ResourceSpec] struct {
+	Count int              `mapstructure:"count"`
+	Items []map[string]any `mapstructure:"items"`
+}
+
+type Resource[T ResourceSpec] struct {
+	Type string `mapstructure:"type"`
+	Spec T      `mapstructure:"spec"`
+}
+
+type ResourceHandler[T ResourceSpec] interface {
+	SetID(T, id.ID) T
+	List(_ context.Context, take, skip int, query, sortBy, sortDirection string) ([]T, error)
+	Count(_ context.Context, query string) (int, error)
+	Create(context.Context, T) (T, error)
+	Update(context.Context, T) (T, error)
+	Get(context.Context, id.ID) (T, error)
+	Delete(context.Context, id.ID) error
+}
+
+type manager[T ResourceSpec] struct {
+	resourceType string
+	handler      ResourceHandler[T]
+	idgen        func() id.ID
+}
+
+func New[T ResourceSpec](resourceType string, handler ResourceHandler[T], idgenFn func() id.ID) *manager[T] {
+	return &manager[T]{
+		resourceType: resourceType,
+		handler:      handler,
+		idgen:        idgenFn,
+	}
+}
+
+func (m *manager[T]) RegisterRoutes(r *mux.Router) *mux.Router {
+	// prefix is /{resourceType | lowercase}/
+	subrouter := r.PathPrefix("/" + strings.ToLower(m.resourceType)).Subrouter()
+
+	subrouter.HandleFunc("/", m.list).Methods(http.MethodGet).Name(fmt.Sprintf("%s.List", m.resourceType))
+	subrouter.HandleFunc("/", m.create).Methods(http.MethodPost).Name(fmt.Sprintf("%s.Create", m.resourceType))
+	subrouter.HandleFunc("/", m.update).Methods(http.MethodPut).Name(fmt.Sprintf("%s.Update", m.resourceType))
+
+	subrouter.HandleFunc("/{id}", m.get).Methods(http.MethodGet).Name(fmt.Sprintf("%s.Get", m.resourceType))
+	subrouter.HandleFunc("/{id}", m.delete).Methods(http.MethodDelete).Name(fmt.Sprintf("%s.Delete", m.resourceType))
+
+	return subrouter
+}
+
+func (m *manager[T]) create(w http.ResponseWriter, r *http.Request) {
```

Comment:
These functions are the real essence of this PR. This code is very much improvable, but provides the basic functionality